File: market_types.cpp
#include "market_types.hpp"
#include <format>
#include <stdexcept>

Price::Price(std::string_view price_string)
{
    size_t decimal_point{price_string.find('.')};
    if (decimal_point == std::string_view::npos || decimal_point != price_string.size() - 3)
    {
        throw std::invalid_argument(
            std::format("Price string without decimal point: {}", price_string));
    }

    std::int64_t price{};
    for (size_t i{}; i < decimal_point; i++)
    {
        if ('0' > price_string[i] || price_string[i] > '9')
        {
            throw std::invalid_argument("Price string is not all digits");
        }

        price *= 10;
        price += (price_string[i] - '0');
    }
    price *= 10'000;

    std::int64_t cents{};
    for (size_t i{decimal_point + 1}; i < price_string.size(); i++)
    {
        if ('0' > price_string[i] || price_string[i] > '9')
        {
            throw std::invalid_argument("Price string is not all digits");
        }

        cents *= 10;
        cents += (price_string[i] - '0');
    }

    cents *= 100;
    price += cents;
    m_price = price;
}

File: market_types.py
from enum import Enum


class Symbol(Enum):
    SYNTH1 = "SYNTH1"
    SYNTH2 = "SYNTH2"
    SYNTH3 = "SYNTH3"
    SYNTH4 = "SYNTH4"
    SYNTH5 = "SYNTH5"


class Aggressor(Enum):
    BUY = "B"
    SELL = "S"


class Price:
    """Abstracts price so that the underlying representation can be changed according to
    dataset requirements.
    """

    __slots__ = ("m_price",)

    def __init__(self, price_string: str):
        """Parses a price string of the form "DD.CC", e.g. "87.54".

        Raises:
            ValueError: if the input is malformed.
        """
        decimal_point = price_string.find(".")
        if decimal_point == -1 or decimal_point != len(price_string) - 3:
            raise ValueError(
                f"Price string without decimal point: {price_string}")

        price = 0
        for ch in price_string[:decimal_point]:
            if not ch.isdigit():
                raise ValueError("Price string is not all digits")
            price = price * 10 + int(ch)
        price *= 10_000

        cents = 0
        for ch in price_string[decimal_point + 1:]:
            if not ch.isdigit():
                raise ValueError("Price string is not all digits")
            cents = cents * 10 + int(ch)
        cents *= 100

        self.m_price = price + cents
